projects/cm_mcu/I2CCommunication.c

```diff
@@ -143,24 +147,27 @@ int apollo_i2c_ctl_reg_r(uint8_t device, uint8_t address, uint8_t reg_address, u
     return *p_status;
 }
 
-int apollo_i2c_ctl_reg_w(uint8_t device, uint8_t address, uint8_t reg_address, uint8_t nbytes, int packed_data)
+int apollo_i2c_ctl_reg_w(uint8_t device, uint8_t address, uint8_t nbytes_addr, uint16_t packed_reg_address, uint8_t nbytes, uint32_t packed_data)
 {
   tSMBus* p_sMaster = pSMBus[device];
   tSMBusStatus* p_eStatus = eStatus[device];
 
   configASSERT(p_sMaster != NULL);
 
-  // first byte is the register, others are the data
-  uint8_t data[MAX_BYTES+1];
-  data[0] = reg_address;
+  // first byte (if write to one of five clcok chips) or two bytes (if write to EEPROM) is the register, others are the data
+  uint8_t data[MAX_BYTES_ADDR + MAX_BYTES];
+  for (int i = 0; i < MAX_BYTES_ADDR; ++i){
+	data[i] = (packed_reg_address >> (nbytes_addr - 1 - i) * 8) & 0xFFUL;
+    if (data[i] != 0) ++nbytes; // to account for the register address
+  }
   // pack the bytes into the data array, offset by
-  // one due to the address
-  for (int i = 1; i < MAX_BYTES + 1; ++i) {
-    data[i] = (packed_data >> (i - 1) * 8) & 0xFFUL;
+  // one or two due to the address
+  for (int i = MAX_BYTES_ADDR; i < MAX_BYTES + MAX_BYTES_ADDR; ++i) {
```

Comment:
does this do the right thing for single-byte addresses? I think it does not. shouldn't the counter be to `nbytes_addr`?

Reply:
I happen to assign data array as data[MAX_BYTES + MAX_BYTES_ADDR]. So, for a one-byte address, the data[1] is empty. Should that be okay?

Reply:
I think this effectively sends a two byte address to devices that do not expect (and support) it....

Reply:
Okay. Your suggestion makes sense. I'll fix it right now.